### PR TITLE
upload and delete build tasks, list --all, error handling, some refactoring

### PR DESCRIPTION
**Pull request by teddyward**
opened a new pull request because the old one was getting ugly with line number changes and there are more major changes now.  I think I've dealt with the error handling issues from the first pull request.  Also added the list --all option to build-tasks-list and tried to un-hardcode argument names / formatting out to make commands more consistent.

I can probably refactor the individual exec() methods a lot but wanted to make sure the overall structure was okay first.  Some refactoring will require changing controller methods -- eg I think we want listing only the newest version of build tasks to be an option on the getTaskDefinitions() server operation.

Thanks for the feedback on the old PR... hoping this one is a bit more complete

### CODE REVIEW

File: app/exec/build-list.ts [hunk 1]
```diff
@@ -2,9 +2,17 @@ import cmdm = require('../lib/tfcommand');
 import cm = require('../lib/common');
 import buildifm = require('vso-node-api/interfaces/BuildInterfaces');
 import buildm = require('vso-node-api/BuildApi');
+import parameternames = require('../lib/parameternames');
```
Comment (bryanmacfarlane) [inline app/exec/build-list.ts]: nit - change to 'params' - makes code downstream not so verbose

File: app/exec/build-tasks-upload.ts [hunk 2]
```diff
@@ -1,17 +1,25 @@
 /// <reference path="../../definitions/q/Q.d.ts"/>
+/// <reference path="../../definitions/node/node.d.ts"/>
 
 import cmdm = require('../lib/tfcommand');
 import cm = require('../lib/common');
+import validatem = require('../lib/jsonvalidate')
```
Comment (bryanmacfarlane) [inline app/exec/build-tasks-upload.ts]: change to vm

File: app/lib/jsonvalidate.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+var fs = require('fs');
+var shell = require('shelljs');
+var check = require('validator');
+
+export class JsonValidate {
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: TaskJsonValidator

File: app/lib/jsonvalidate.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	/*
+	 * Checks a json file for correct formatting against some validation function
+	 * @param jsonFilePath path to the json file being validated
+	 * @param jsonValidationFunction function that validates parsed json data against some criteria
+	 * @return the parsed json file
+	 * @throws InvalidDirectoryException if json file doesn't exist, InvalidJsonException on failed parse or *first* invalid field in json
+	*/
+	public validateJson(jsonFilePath: string, jsonValidationFunction: (path: string, jsonData: any) => void, jsonMissingErrorMessage?: string): any {
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: validate

File: app/lib/jsonvalidate.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		var jsonMissingErrorMsg: string = jsonMissingErrorMessage || 'specified json file does not exist.'
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: tab spacing

File: app/lib/jsonvalidate.ts (hunk 4 above)
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: why the level of indirection with the jsonValidationFunction?  For that matter, why the class?  This module could just export one function - validateTask

File: app/lib/jsonvalidate.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        this.exists(jsonFilePath, jsonMissingErrorMsg);
+
+        var taskJson;
+        try {
+            taskJson = JSON.parse(fs.readFileSync(jsonFilePath, 'utf8'));
+        }
+        catch (jsonError) {
+            throw new InvalidJsonException("Invalid task json: " + jsonError);
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: Let's make this function return a Q.Promise<void>.  An error would reject with a good message

File: app/lib/jsonvalidate.ts (hunk 4 above)
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: I'm fine with the validation being another function, just wierd that the caller passes it.  It should internally call it.

File: app/lib/jsonvalidate.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            throw new InvalidFieldException(vn + ': friendlyName is a required string <= 40 chars');
+        }
+
+        if (!taskData.instanceNameFormat) {
+            throw new InvalidFieldException(vn + ': instanceNameFormat is required');
+        }
+	}
+}
+
+export class InvalidJsonException implements Error {
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: We don't need all of these exception classes.  reject(new Error('...')) is fine

File: app/lib/jsonvalidate.ts (hunk 4 above)
Comment (teddyward) [inline app/lib/jsonvalidate.ts]: I guess I was thinking that there could be other non-task commands that require json validation but maybe that's unrealistic. I'll get rid of the class

File: app/lib/jsonvalidate.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+	/*
+     * Validates a parsed json file describing a build task
+     * @param taskPath the path to the original json file
+     * @param taskData the parsed json file
+     */
+	public validateTask(taskPath: string, taskData: any) {
+        var vn = (taskData.name || taskPath);
+
+        if (!taskData.id || !check.isUUID(taskData.id)) {
```
Comment (bryanmacfarlane) [inline app/lib/jsonvalidate.ts]: Would be better to have this function return a string[] list of issues and not throw so we could get a collection of all the issues in one output

File: app/lib/loader.ts [hunk 9]
```diff
@@ -107,14 +107,12 @@ export function getHelp(execPath: string, scope: string, all: boolean) {
...
-                var args = mod.argsFormat ? mod.argsFormat() : '';
-                args = args.length > 0 ? ' ' + args : args;
                 var description = mod.describe ? mod.describe() : '';
-                var optionsLabel = hasImplementation ? ' [options]' : '';
+                var listedArguments = hasImplementation ? mod.getArguments() : '';
                 console.log(colors.yellow('   ' + cmdLabel));
                 console.log(colors.white('\t' + description));
                 if (hasImplementation) 
-                    console.log('\t' + cmdLabel + args + optionsLabel);
+                    console.log('\t' + cmdLabel + listedArguments);
```
Comment (bryanmacfarlane) [inline app/lib/loader.ts]: would like to see what this looks like (screenshot)?

File: app/lib/parameternames.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+export var ALL: string = 'all';
```
Comment (bryanmacfarlane) [inline app/lib/parameternames.ts]: this is good to drive consistency.  thanks

File: app/lib/loader.ts (hunk 9 above)
Comment (teddyward) [inline app/lib/loader.ts]: arghelp

File: app/lib/jsonvalidate.ts (hunk 8 above)
Comment (teddyward) [inline app/lib/jsonvalidate.ts]: still should throw eventually though right?  or should I let users upload an incomplete task and just print these as warnings?

File: app/exec/build-tasks-delete.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+export class BuildTaskDelete extends cmdm.TfCommand {
+    public exec(args: string[], options: cm.IOptions): any {
+        var taskId = args[0] || options[parameternames.GENERIC_ID];
+        this.checkRequiredParameter(taskId, parameternames.GENERIC_ID, parameternames.TASK_ID);
+        var deferred = Q.defer<agentifm.TaskDefinition>();
+
+        var agentapi = this.getWebApi().getTaskAgentApi(this.connection.accountUrl);
+
+        //TODO: update server api to give ability to check if task exists first
```
Comment (teddyward) [inline app/exec/build-tasks-delete.ts]: er - nevermind, just going to try a getTaskContent here.  slower, but doesn't require changing controllers

**Pull request merged**